shared/domains/chelonia/internals.js

```diff
@@ -1,19 +1,86 @@
 import { ChelErrorUnexpected, ChelErrorUnrecoverable } from './errors.js'
 import { CONTRACT_IS_SYNCING, CONTRACTS_MODIFIED, EVENT_HANDLED } from './events.js'
+import { handleFetchResult } from '~/frontend/controller/utils/misc.js'
+import { blake32Hash } from '~/shared/functions.js'
+// import 'ses'
 
 import type { GIOpContract, GIOpType, GIOpActionEncrypted, GIOpActionUnencrypted, GIOpPropSet, GIOpKeyAdd } from './GIMessage.js'
 
-sbp('sbp/selectors/register', {
+export default (sbp('sbp/selectors/register', {
   //     DO NOT CALL ANY OF THESE YOURSELF!
   'chelonia/private/state': function () {
     return this.state
   },
+  'chelonia/private/loadManifest': async function (manifestHash: string) {
+    if (this.manifestToContract[manifestHash]) {
+      console.warn('[chelonia]: already loaded manifest', manifestHash)
+      return
+    }
+    const manifestURL = `${this.config.connectionURL}/file/${manifestHash}`
+    const manifest = await fetch(manifestURL).then(handleFetchResult('json'))
+    const body = JSON.parse(manifest.body)
+    const contractInfo = (this.config.contracts.defaults.preferSlim && body.contractSlim) || body.contract
+    console.debug('[chelonia] loading contract', contractInfo.file, 'from manifest:', manifestHash)
+    const source = await fetch(`${this.config.connectionURL}/file/${contractInfo.hash}`)
+      .then(handleFetchResult('text'))
+    const sourceHash = blake32Hash(source)
+    if (sourceHash !== contractInfo.hash) {
+      throw new Error(`bad hash ${sourceHash} for contract '${contractInfo.file}'! Should be: ${contractInfo.hash}`)
+    }
+    function reduceAllow (acc, v) { acc[v] = true; return acc }
+    const allowedSels = ['okTurtles.events/on', 'chelonia/defineContract']
+      .concat(this.config.contracts.defaults.allowedSelectors)
+      .reduce(reduceAllow, {})
+    const allowedDoms = this.config.contracts.defaults.allowedDomains
+      .reduce(reduceAllow, {})
+    let contractName: string // eslint-disable-line prefer-const
+    const contractSBP = (selector: string, ...args) => {
+      const domain = domainFromSelector(selector)
+      if (selector.startsWith(contractName)) {
+        selector = `${manifestHash}/${selector}`
+      }
+      if (allowedSels[selector] || allowedDoms[domain]) {
+        return sbp(selector, ...args)
+      } else {
+        throw new Error(`[chelonia] selector not on allowlist: '${selector}'`)
+      }
+    }
+    // eslint-disable-next-line no-new-func
+    const saferEval: Function = new Function(`
+      return function (require, globals) {
+        with (globals) {
+          ${source}
+        }
+      }
+    `)()
+    this.defContractSBP = contractSBP
+    this.defContractManifest = manifestHash
+    saferEval((dep) => {
+      return dep === '@sbp/sbp'
+        ? contractSBP
+        : this.config.contracts.defaults.modules[dep]
+    }, {
+      fetch: null,
+      XMLHttpRequest: null,
```

Comment:
Why are we using `XMLHttpRequest`? It seems quite redundant when we're already using `fetch` and all modern browsers support it. I doubt the code otherwise works in browsers that _don't_ support `fetch`, so this seems like a good opportunity to remove it, maybe?

Reply:
We aren't using it, we are making sure the contract cannot use it. This is a lightweight sandbox that undefines global variables we don't want the contract to have access to.

Soon I will push a commit that changes this section to look more like this:

```js
    saferEval((dep) => {
      return dep === '@sbp/sbp'
        ? contractSBP
        : this.config.contracts.defaults.modules[dep]
    }, {
      // not the greatest of sandboxes but something is better than nothing
      window: undefined,
      globalThis: undefined,
      fetch: undefined,
      XMLHttpRequest: undefined,
      // preserve these however, they're useful
      setTimeout,
      clearTimeout,
      setInterval,
      clearInterval,
      // TODO: add other possible global overrides here.
      //       Although again, the contracts are signed so not that big of a deal
      ...this.config.contracts.defaults.exposedGlobals,
      sbp: contractSBP
    })
```